Improved file operations

kamphaus opened the pull request:
Also fixes #90 for `tmpnam()`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/elliotchance/c2go/733)
<!-- Reviewable:end -->

elliotchance commented:
Please fix merge conflicts

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/elliotchance/c2go/733#-:-LCfZltg54lkyDN_KzcJ:bp1lw2l)*
<!-- Sent from Reviewable.io -->

kamphaus commented:
Merge conflicts fixed.

I also added tests and fixed a bug that I hadn't noticed until I added the tests.
That will teach me to write the tests earlier.

elliotchance commented:
Reviewed 1 of 2 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/elliotchance/c2go/733)*
<!-- Sent from Reviewable.io -->
